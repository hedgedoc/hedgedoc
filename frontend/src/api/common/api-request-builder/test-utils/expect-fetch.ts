/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { defaultConfig } from '../../default-config'
import { Mock } from 'ts-mockery'

/**
 * Mock fetch api for tests.
 * Check that the given url and options are present in the request and return the given status code.
 *
 * @param expectedUrl the url that should be requested
 * @param requestStatusCode the status code the mocked request should return
 * @param expectedOptions additional options
 */
export const expectFetch = (
  expectedUrl: string,
  requestStatusCode: number,
  expectedOptions: RequestInit,
  responseBody?: unknown
): void => {
  global.fetch = jest.fn((fetchUrl: RequestInfo | URL, fetchOptions?: RequestInit): Promise<Response> => {
    expect(fetchUrl).toEqual(expectedUrl)
    expect(fetchOptions).toStrictEqual({
      ...defaultConfig,
      body: undefined,
      headers: new Headers(),
      ...expectedOptions
    })
    return Promise.resolve(
      Mock.of<Response>({
        status: requestStatusCode,
        statusText: mapCodeToText(requestStatusCode),
        json: jest.fn(() => (responseBody ? Promise.resolve(responseBody) : Promise.reject()))
      })
    )
  }) as typeof global.fetch
}
const mapCodeToText = (code: number): string => {
  switch (code) {
    case 400:
      return 'bad_request'
    case 401:
      return 'forbidden'
    default:
      return 'unknown_code'
  }
}
