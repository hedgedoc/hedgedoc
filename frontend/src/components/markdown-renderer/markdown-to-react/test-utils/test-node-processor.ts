/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { Document } from 'domhandler'
import { Element, Text } from 'domhandler'

export class TestNodeProcessor extends NodeProcessor {
  process(document: Document): Document {
    document.childNodes.push(new Element('node-processor', {}, [new Text('node processor children')]))
    return document
  }
}
