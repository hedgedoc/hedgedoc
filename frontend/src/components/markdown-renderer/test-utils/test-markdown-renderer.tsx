/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { StoreProvider } from '../../../redux/store-provider'
import type { MarkdownRendererExtension } from '../extensions/_base-classes/markdown-renderer-extension'
import { MarkdownToReact } from '../markdown-to-react/markdown-to-react'
import React, { useMemo } from 'react'

export interface SimpleMarkdownRendererProps {
  content: string
  extensions: MarkdownRendererExtension[]
}

/**
 * A markdown renderer for tests.
 *
 * @param content The content to be rendered.
 * @param extensions The {@link MarkdownRendererExtension MarkdownExtensions} to use for rendering.
 */
export const TestMarkdownRenderer: React.FC<SimpleMarkdownRendererProps> = ({ content, extensions }) => {
  const lines = useMemo(() => content.split('\n'), [content])
  return (
    <StoreProvider>
      <MarkdownToReact
        markdownContentLines={lines}
        markdownRenderExtensions={extensions}
        newlinesAreBreaks={true}
        allowHtml={false}
      />
    </StoreProvider>
  )
}
