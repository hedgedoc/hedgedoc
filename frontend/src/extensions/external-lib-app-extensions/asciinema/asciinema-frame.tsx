/*
 * SPDX-FileCopyrightText: 2024 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { ClickShield } from '../../../components/markdown-renderer/replace-components/click-shield/click-shield'
import type { IdProps } from '../../../components/markdown-renderer/replace-components/custom-tag-with-id-component-replacer'
import React from 'react'
import { Play as IconPlay } from 'react-bootstrap-icons'

/**
 * Renders an embedding for https://asciinema.org
 *
 * @param id The id from the asciinema url
 */
export const AsciinemaFrame: React.FC<IdProps> = ({ id }) => {
  return (
    <ClickShield
      hoverIcon={IconPlay}
      targetDescription={'asciinema'}
      fallbackPreviewImageUrl={`https://asciinema.org/a/${id}.png`}
      fallbackBackgroundColor={'#d40000'}
      containerClassName={''}
      fallbackLink={`https://asciinema.org/a/${id}`}
      data-cypress-id={'click-shield-asciinema'}>
      <span className={'ratio ratio-16x9 d-print-none'}>
        <iframe
          allowFullScreen={true}
          className=''
          title={`asciinema cast ${id}`}
          src={`https://asciinema.org/a/${id}/iframe?autoplay=1`}
        />
      </span>
    </ClickShield>
  )
}
