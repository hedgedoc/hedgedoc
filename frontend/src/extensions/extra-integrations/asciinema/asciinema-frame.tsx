/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { ClickShield } from '../../../components/markdown-renderer/replace-components/click-shield/click-shield'
import type { IdProps } from '../../../components/markdown-renderer/replace-components/custom-tag-with-id-component-replacer'
import React from 'react'
import { Play as IconPlay } from 'react-bootstrap-icons'

/**
 * Renders an embedding for https://asciinema.org
 *
 * @param id The id from the asciinema url
 */
export const AsciinemaFrame: React.FC<IdProps> = ({ id }) => {
  return (
    <ClickShield
      hoverIcon={IconPlay}
      targetDescription={'asciinema'}
      fallbackPreviewImageUrl={`https://asciinema.org/a/${id}.png`}
      fallbackBackgroundColor={'#d40000'}
      containerClassName={''}
      data-cypress-id={'click-shield-asciinema'}>
      <span className={'ratio ratio-16x9'}>
        <iframe
          allowFullScreen={true}
          className=''
          title={`asciinema cast ${id}`}
          src={`https://asciinema.org/a/${id}/iframe?autoplay=1`}
        />
      </span>
    </ClickShield>
  )
}
