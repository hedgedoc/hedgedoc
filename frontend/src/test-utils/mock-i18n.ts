/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { TFunction } from 'i18next'
import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'

/**
 * Initializes i18n with minimal settings and without any data, so it just returns the used key as translation.
 *
 * @return A promise that resolves if i18n has been initialized
 */
export const mockI18n = (): Promise<TFunction> => {
  return i18n.use(initReactI18next).init({
    lng: 'en',
    fallbackLng: 'en',
    ns: ['translationsNS'],
    defaultNS: 'translationsNS',
    interpolation: {
      escapeValue: false
    },
    resources: { en: { translationsNS: {} } }
  })
}
