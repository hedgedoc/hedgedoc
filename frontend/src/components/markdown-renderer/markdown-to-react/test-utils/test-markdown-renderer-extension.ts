/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { MarkdownRendererExtension } from '../../extensions/_base-classes/markdown-renderer-extension'
import type { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { ComponentReplacer } from '../../replace-components/component-replacer'
import { TestNodeProcessor } from './test-node-processor'
import { TestReplacer } from './test-replacer'
import type MarkdownIt from 'markdown-it'
import Token from 'markdown-it/lib/token'

export class TestMarkdownRendererExtension extends MarkdownRendererExtension {
  buildNodeProcessors(): NodeProcessor[] {
    return [new TestNodeProcessor()]
  }

  configureMarkdownIt(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('configure', (core) => core.tokens.push(new Token('configure', 'configure', 0)))
      markdownIt.renderer.rules.configure = () => '<span>configure</span>'
    })
  }

  buildReplacers(): ComponentReplacer[] {
    return [new TestReplacer()]
  }

  buildTagNameAllowList(): string[] {
    return ['node-processor']
  }

  configureMarkdownItPost(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('post', (core) => core.tokens.push(new Token('post', 'post', 0)))
      markdownIt.renderer.rules.post = () => '<span>post</span>'
    })
  }
}
