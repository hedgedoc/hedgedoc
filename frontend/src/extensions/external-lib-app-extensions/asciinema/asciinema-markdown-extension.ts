/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/_base-classes/markdown-renderer-extension'
import type { ComponentReplacer } from '../../../components/markdown-renderer/replace-components/component-replacer'
import { CustomTagWithIdComponentReplacer } from '../../../components/markdown-renderer/replace-components/custom-tag-with-id-component-replacer'
import { AsciinemaFrame } from './asciinema-frame'
import { replaceAsciinemaLinkMarkdownItPlugin } from './replace-asciinema-link'
import type MarkdownIt from 'markdown-it/lib'

/**
 * Adds asciinema embeddings to the markdown rendering by detecting asciinema.org links.
 */
export class AsciinemaMarkdownExtension extends MarkdownRendererExtension {
  public static readonly tagName = 'app-asciinema'

  public configureMarkdownIt(markdownIt: MarkdownIt): void {
    replaceAsciinemaLinkMarkdownItPlugin(markdownIt)
  }

  public buildReplacers(): ComponentReplacer[] {
    return [new CustomTagWithIdComponentReplacer(AsciinemaFrame, AsciinemaMarkdownExtension.tagName)]
  }

  public buildTagNameAllowList(): string[] {
    return [AsciinemaMarkdownExtension.tagName]
  }
}
