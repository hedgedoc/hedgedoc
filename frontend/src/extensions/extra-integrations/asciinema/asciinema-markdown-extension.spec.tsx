/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { IdProps } from '../../../components/markdown-renderer/replace-components/custom-tag-with-id-component-replacer'
import { TestMarkdownRenderer } from '../../../components/markdown-renderer/test-utils/test-markdown-renderer'
import { mockI18n } from '../../../test-utils/mock-i18n'
import * as AsciinemaFrameModule from './asciinema-frame'
import { AsciinemaMarkdownExtension } from './asciinema-markdown-extension'
import { render } from '@testing-library/react'
import React from 'react'

jest.mock('./asciinema-frame')

describe('asciinema markdown extension', () => {
  beforeAll(async () => {
    jest
      .spyOn(AsciinemaFrameModule, 'AsciinemaFrame')
      .mockImplementation((({ id }) => (
        <span>this is a mock for the asciinema frame with id {id}</span>
      )) as React.FC<IdProps>)
    await mockI18n()
  })

  afterAll(() => {
    jest.resetAllMocks()
    jest.resetModules()
  })

  it('renders plain asciinema URLs', () => {
    const view = render(
      <TestMarkdownRenderer
        extensions={[new AsciinemaMarkdownExtension()]}
        content={'https://asciinema.org/a/190123709'}
      />
    )
    expect(view.container).toMatchSnapshot()
  })
})
