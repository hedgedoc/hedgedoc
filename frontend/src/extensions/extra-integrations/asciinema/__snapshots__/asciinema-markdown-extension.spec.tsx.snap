// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asciinema markdown extension renders plain asciinema URLs 1`] = `
<div>
  <p>
    <span>
      this is a mock for the asciinema frame with id 
      190123709
    </span>
  </p>
  

</div>
`;
